src/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -1279,8 +1244,200 @@ internal bool VerifyRemoteCertificate(RemoteCertValidationCallback remoteCertVal
+            }
+
+            return token;
+        }
+
+        public ProtocolToken CreateShutdownToken()
+        {
+            if (GlobalLog.IsEnabled)
+            {
+                GlobalLog.Enter("SecureChannel#" + LoggingHash.HashString(this) + "::CreateShutdownToken");
+            }
+
+            SecurityStatusPal status;
+            status = SslStreamPal.ApplyShutdownToken(ref _credentialsHandle, _securityContext);
+
+            if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+            {
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Print("SecureChannel#" + LoggingHash.HashString(this) + "::ApplyAlertToken() returned " + status.ErrorCode);
+                }
+
+                throw status.Exception;
+            }
+
+            ProtocolToken token = GenerateAlertToken();
+
+            if (GlobalLog.IsEnabled)
+            {
+                GlobalLog.Leave("SecureChannel#" + LoggingHash.HashString(this) + "::CreateShutdownToken", token.ToString());
+            }
+
+            return token;
+        }
+
+        private ProtocolToken GenerateAlertToken()
+        {
+            byte[] nextmsg = null;
+
+            SecurityStatusPal status;
+            status = GenerateToken(null, 0, 0, ref nextmsg);
+
+            ProtocolToken token = new ProtocolToken(nextmsg, status);
+
+            return token;
+        }
+        
+        private static TlsAlertMessage GetAlertMessageFromChain(X509Chain chain)
+        {
+            foreach (X509ChainStatus chainStatus in chain.ChainStatus)
+            {
+                if (chainStatus.Status == X509ChainStatusFlags.NoError)
+                {
+                    continue;
+                }
+
+                if ((chainStatus.Status &
+                    (X509ChainStatusFlags.UntrustedRoot | X509ChainStatusFlags.PartialChain |
+                     X509ChainStatusFlags.Cyclic)) != 0)
+                {
+                    return TlsAlertMessage.UnknownCA;
+                }
+
+                if ((chainStatus.Status &
+                    (X509ChainStatusFlags.Revoked | X509ChainStatusFlags.OfflineRevocation |
+                     X509ChainStatusFlags.RevocationStatusUnknown)) != 0)
+                {
+                    return TlsAlertMessage.CertificateRevoked;
+                }
+
+                if ((chainStatus.Status &
+                    (X509ChainStatusFlags.CtlNotTimeValid | X509ChainStatusFlags.NotTimeNested |
+                     X509ChainStatusFlags.NotTimeValid)) != 0)
+                {
+                    return TlsAlertMessage.CertificateExpired;
+                }
+
+                if ((chainStatus.Status &
+                    (X509ChainStatusFlags.NotValidForUsage | X509ChainStatusFlags.CtlNotValidForUsage)) != 0)
+                {
+                    return TlsAlertMessage.UnsupportedCert; 
```

Comment:
NotValidForUsage means "you told me it needed to be a TlsServerAuthentication certificate, but it wasn't".  I think that's more BadCertificate than Unsupported.  My reading of Unsupported is more "I don't know how to read a v4 certificate"